pkg/labels/labels.go

```diff
@@ -2,24 +2,146 @@ package labels
 
 import (
 	"errors"
+	"regexp"
+	"strings"
+	"unicode"
 
-	"k8s.io/apimachinery/pkg/labels"
+	dfutil "github.com/devfile/library/pkg/util"
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/klog"
 
 	"github.com/redhat-developer/odo/pkg/version"
 )
 
+var _replacementMap = map[string]string{
+	".": "dot",
+	"#": "sharp",
+}
+
+var (
+	_regexpInvalidCharacters        = regexp.MustCompile(`[^a-zA-Z0-9._-]+`)
+	_regexpStartingWithAlphanumeric = regexp.MustCompile(`^[a-z0-9A-Z]`)
+	_regexpEndingWithAlphanumeric   = regexp.MustCompile(`[a-z0-9A-Z]$`)
+)
+
 // GetLabels return labels that should be applied to every object for given component in active application
 // if you need labels to filter component then use GetSelector instead
 // Note: isPartOfComponent denotes if the label is required for a core resource(deployment, svc, pvc, pv) of a given component deployed with `odo dev`;
 // it is the only thing that sets it apart from the resources created via other ways (`odo deploy`, deploying resource with apply command during `odo dev`)
 func GetLabels(componentName string, applicationName string, runtime string, mode string, isPartOfComponent bool) map[string]string {
 	labels := getLabels(componentName, applicationName, mode, true, isPartOfComponent)
 	if runtime != "" {
-		labels[openshiftRunTimeLabel] = runtime
+		labels[openshiftRunTimeLabel] = sanitizeLabelValue(openshiftRunTimeLabel, runtime)
 	}
 	return labels
 }
 
+// sanitizeLabelValue makes sure that the value v specified is a valid value for a Kubernetes label,
+// by sanitizing it.
+//
+// Valid label value:
+//
+// - must be 63 characters or fewer (can be empty),
+//
+// - unless empty, must begin and end with an alphanumeric character ([a-z0-9A-Z]),
+//
+// - could contain dashes (-), underscores (_), dots (.), and alphanumerics between.
+//
+// As such, sanitizeLabelValue might:
+//
+// - truncate the label value so that it is less than 63 characters,
```

Comment:
You can move this point after the replcments, as it is done after. Also, you could add a test that checks that a long string with . or # is truncated to 63 after replacement (and not before).

Reply:
Done in [`d6941ee` (#6284)](https://github.com/redhat-developer/odo/pull/6284/commits/d6941eeb792bf30a338ead7f76de61ac701dbd41) - thanks for pointing this out.